engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java

```diff
@@ -111,23 +113,35 @@ private void loadModulesFromApplicationPath(PathManager pathManager) {
     }
 
     private Module loadEngineModule(List<Class<?>> classesOnClasspathsToAddToEngine) {
-        // FIXME: is `classes…toAddToEngine` gone? Did we ever use it in the first place?
-        Module engine = moduleFactory.createPackageModule("org.terasology.engine");
-        // TODO: document why nui is a module when other libraries are not
-        Module nui = createSyntheticPackageModule("nui", "3.0.0" /* FIXME */, "org.terasology.nui");
-
-        DependencyInfo nuiDependency = new DependencyInfo();
-        nuiDependency.setId(nui.getId());
-        nuiDependency.setMinVersion(nui.getVersion());
-        nuiDependency.setMaxVersion(nui.getVersion().getNextPatchVersion());
-        engine.getMetadata().getDependencies().add(nuiDependency);
-
-        registry.add(nui);
-        registry.add(engine);
+        Module packageModule = moduleFactory.createPackageModule("org.terasology.engine");
+
+        // We need to add reflections from our subsystems and other classes.
+        ConfigurationBuilder config = new ConfigurationBuilder();
+        Reflections packageReflections = packageModule.getModuleManifest();
+        Set<Scanner> scanners = packageReflections.getConfiguration().getScanners();
+        config.setScanners(scanners.toArray(new Scanner[0]));
+
+        for (Class<?> aClass : classesOnClasspathsToAddToEngine) {
+            config.addUrls(ClasspathHelper.forClass(aClass));
+        }
+
+        // TODO: is this using reflections.cache?
+        Reflections reflectionsWithSubsystems = new Reflections(config);
+        packageReflections.merge(reflectionsWithSubsystems);
 
-        // FIXME: why doesn't gestalt-v7 need this?
-        //  enrichReflectionsWithSubsystems(engineModule);
+        ClassLoader engineClassLoader = this.getClass().getClassLoader();
 
+        // We need the class predicate to include classes in subsystems and whatnot. We can't change it in an
+        // existing module, so make a new one based on the one from the moduleFactory.
+        Module engine = new Module(
+                packageModule.getMetadata(),
+                packageModule.getResources(),
+                Collections.emptyList(),
+                packageReflections,
+                packageModule.getClassPredicate().or(clazz -> clazz.getClassLoader().equals(engineClassLoader))
+        );
```

Comment:
This whole method is essentially doing gestalt ModuleFactory stuff. A sign that we might want to add a new method there or expose some of its private methods.